File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/EmoteCommand.vb
Imports AtlasV.Localization
Imports System.Collections.Generic

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class EmoteCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            If varContext.GameState.ActiveChannel Is Nothing Then
                Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, CUInt(0), varContext.GameState.Ping, varContext.GameState.OnlineName, Resources.InvalidChatCommand).WriteTo(varContext.GameState.Client)
                Return
            End If

            varContext.GameState.ActiveChannel.WriteChatMessage(varContext.GameState, RawBuffer, True)

            If varContext.GameState.ActiveChannel.Count <= 1 OrElse varContext.GameState.ActiveChannel.ActiveFlags.HasFlag(Channel.Flags.Silent) Then
                Call New ChatEvent(ChatEvent.EventIds.EID_INFO, varContext.GameState.ActiveChannel.ActiveFlags, 0, varContext.GameState.ActiveChannel.Name, Resources.NoOneHearsYou).WriteTo(varContext.GameState.Client)
            End If
        End Sub
    End Class
End Namespace

File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/EmoteCommand.cs
using Atlasd.Localization;
using System.Collections.Generic;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class EmoteCommand : ChatCommand
    {
        public EmoteCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            if (context.GameState.ActiveChannel == null)
            {
                new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, Resources.InvalidChatCommand).WriteTo(context.GameState.Client);
                return;
            }

            context.GameState.ActiveChannel.WriteChatMessage(context.GameState, RawBuffer, true);

            if (context.GameState.ActiveChannel.Count <= 1 || context.GameState.ActiveChannel.ActiveFlags.HasFlag(Channel.Flags.Silent))
            {
                new ChatEvent(ChatEvent.EventIds.EID_INFO, context.GameState.ActiveChannel.ActiveFlags, 0, context.GameState.ActiveChannel.Name, Resources.NoOneHearsYou).WriteTo(context.GameState.Client);
            }
        }
    }
}
